Benchmark vignette

jeffreyhanson opened the pull request:
This PR is to add the benchmark vignette. Since the benchmark vignette is not finished, this PR should not be merged. @ricschuster, I've added in the code to download and import the results from the https://github.com/prioritizr/benchmark repo. Can you verify if it works for you?

jeffreyhanson commented:
Assuming that the data import code works, @ricschuster could you please write up the vignette when you get a chance? I'd be happy to look over any draft versions if that would be helpful?

jeffreyhanson commented:
Also, it seems like Gurobi has some really long run times (e.g. see the plot). I wonder if this is because the `time_limit` parameter doesn't include presolve steps?

ricschuster commented:
Thanks Jeff!
Loading of the data works.

It is odd that Gurobi would be having such high run times. It might be related to the pre-solve steps not counting to the run time.

What's even more puzzling to me is that none of the other solvers seem to have long run times. I think something is strange is going on here. Especially for the bigger problem sizes all solvers, but especially the open source ones (other than CBC) should have longer run times than we see in the plot `p`. 

What do you think?

jeffreyhanson commented:
> It is odd that Gurobi would be having such high run times. It might be related to the pre-solve steps not counting to the run time.

Yeah - maybe we should rerun the benchmarks with presolve set to -1 for Gurobi?

> What's even more puzzling to me is that none of the other solvers seem to have long run times. I think something is strange is going on here. Especially for the bigger problem sizes all solvers, but especially the open source ones (other than CBC) should have longer run times than we see in the plot p.

Just to check, what's the time limit set as?

ricschuster commented:
Will set presolve to -1.

Time limit is 1000 seconds.
Open source solvers should use that for the bigger problems I think. Probably good to do some run time testing.

jeffreyhanson commented:
Hmm, I was just checking how we measure run time across the different solvers. It seems like we rely on [Gurobi's internal runtime reporting](https://github.com/prioritizr/prioritizr/blob/master/R/add_gurobi_solver.R#L197-L215), but actually calculate run times manually for all other solvers (e.g. see [CPLEX](https://github.com/prioritizr/prioritizr/blob/master/R/add_cplex_solver.R#L137-L157)) . I wonder if we should update the Gurobi solver to calculate runtimes in R to be consistent?

ricschuster commented:
> 
> 
> Hmm, I was just checking how we measure run time across the different solvers. It seems like we rely on [Gurobi's internal runtime reporting](https://github.com/prioritizr/prioritizr/blob/master/R/add_gurobi_solver.R#L197-L215), but actually calculate run times manually for all other solvers (e.g. see [CPLEX](https://github.com/prioritizr/prioritizr/blob/master/R/add_cplex_solver.R#L137-L157)) . I wonder if we should update the Gurobi solver to calculate runtimes in R to be consistent?

Sounds like a good idea to me.

I'm mainly concerned about the reported timing results and will try out some examples tomorrow, just to make sure we are measuring correctly and that everything works okay. They code indicates it should, but I just don't trust those times that we got.

jeffreyhanson commented:
Ah - just realized that the gurobi timing probably should be in master so we can easily add it tot the benchmark packrat setup. Since this is a really minor change, I'll just make the update and send it to main branch. I'll let you know when I've updated the benchmark repo with the new version.

jeffreyhanson commented:
> I'm mainly concerned about the reported timing results and will try out some examples tomorrow, just to make sure we are measuring correctly and that everything works okay. They code indicates it should, but I just don't trust those times that we got.

Yeah - me as well. Thanks! Let me know if you work out what's going on?

ricschuster commented:
Thanks Jeff!

Will see what I can find out tomorrow and let you know.

jeffreyhanson commented:
Ok, I've updated prioritizr with the standardized run time measurements for Gurobi, and updated the version of prioritizr in the benchmark repo so that it uses the new version of prioritizr. So, I think it's ready for you now? Let me know if you run into any issues.

ricschuster commented:
@jeffreyhanson could you have a quick look at `benchmark.Rmd` and let me know if you can think of any other plots we should include for this vignette?

I think the most informative plots are there and its mostly a matter of filling in the text now.
What do you think?

jeffreyhanson commented:
Sorry for the delay on getting back to you on this. I think this is looking really good! I like how you've kept run time in absolute units -- I think this will be more accessible for most users. I also really like how you've written a function to automate creating plots -- it makes things much easier to understand. In terms of suggestions for improvements, what do you think about:

1. Is it worth looking at 2 different BLM values to show how the BLM value used also influences run time?
2. Adding the overall benchmark results to an "overall" subsection of the results? Perhaps you could have the following subsections in the results: "Min set with no boundary penalty", "Min set with low boundary penalty", and "Min set with high boundary penalty"
3. Are the plots/interpretations taking into accounted the exceeded run time data in the table? E.g. the plots might show that CBC runs very quickly but this might be because it's hitting the time limit?
4. Tweaking the benchmark plot function to have an extra parameter to specify which solvers should be shown in the plot? This could default to all solvers, and for the CBC/CPLEX/Gurobi comparisons you could specify a `character` vector to just show these solvers? Happy to prototype this functionality if I'm not explaining this well? 
5. It might be nice to do some automatic units conversions for the plots, so time values are shown in hours (for example) if that makes more sense than 10,000+ seconds. I'm happy to look into this if that would help?
6. Is it worth including some discussion/recommendations/conclusions section after the results to emphasize the main points?

ricschuster commented:
Thanks very much Jeff!

1. I agree, adding another BLM value would be really useful.
2. Good idea.
3. Not on my desktop right now, but as far as I remember only SYMPHONY based `add_solver` funtions hit the time limit. Might be a good idea to remove the time limit to not run into the potential issue. Given that we only run the benchmark occasionally that's probably okay and would help get a real sense of run time differences.
4. Yeah, I was thinking about that as well, but haven't implemented it yet. If you want to prototype that, that would be really helpful.
5. If you could look into that as well, that would be awesome.
6. Definitely. Was planning on that as I flash out the text.

One additional large component that I thought might be good to add is at least another solver. Would be interesting to see how solvers compare for at least one other commonly used solver. What do you think?

jeffreyhanson commented:
Ok excellent - thanks! I'll try and play around with the plots sometime over the next couple of days and let you know how I go.

ricschuster commented:
Quick follow up: what do you think about expanding the benchmark to include another objective function?

What do you think about `add_min_largest_shortfall_objective`?

If you think that's a good idea, would you mind adjusting `03-analysis.R` to allow for the two objective functions?

jeffreyhanson commented:
Just so we've got this written down somewhere, as discussed during the meeting, I think this is a great idea and I'll add the min shortfall objective.

jeffreyhanson commented:
@ricschuster, absolutely brilliant work! I've tried to editing the text to have consistent formatting with other vignettes, flow a bit better, and emphasize the main points a bit more. Let me know what you think? Please feel free to revert any changes you disagree with. I've also removed a couple of graphs (i.e. the last ones comparing CBC vs CPLEX vs Gurobi) where they don't really add much new information (e.g. because we can see the same results in the graphs comparing all the solvers together).

ricschuster commented:
Thanks very much @jeffreyhanson ! 
Edits look great.
I'm happy for you to write up a conclusion section/paragraph, if you'd like.
I would recommend CBC over lpsymphony and Rsymphony for all cases. It might be a bit too broad a statement based on what we have tested, but I think it would work for most people.

As for Gurobi vs CBC, CBC is doing pretty well for min_set, but if Gurobi is available to users its probably best to always go for that one. Have you ever looked at the license costs for CPLEX? I wonder if CPLEX was cheaper than Gurobi, we could also speak to that at one point?

Anyways, I think this is getting really close, which is awesome!

jeffreyhanson commented:
Ok awesome - sound great! Yeah, I can draft up a conclusion section and add that to the vignette.

Interesting question re CPLEX vs. Gurobi costs, it looks like CPLEX is $200USD per month for a single user license (https://www.ibm.com/products/ilog-cplex-optimization-studio/pricing). Since Gurobi is several thousand USD per year (last time I checked?), I think that means CPLEX is cheaper?

ricschuster commented:
Thanks Jeff!

re: CPLEX. Hard to say what deployed cost would be for them. This $200/month are only for development work. I guess for individual use it would probably be okay, but deploying apps would be interesting to know too. For single users it does seem like a cheaper option that Gurobi for sure. Do you think that's worth mentioning in the vignette too?

jeffreyhanson commented:
@ricschuster, I've just pushed an update which tweaks some of the writing and adds a concluding paragraph. Can you please take a quick look at the paragraph and skim for any typos? I've attached the html doc of the Rmd to this post in case that's helpful? Once you're happy with it, I'll open up a new PR and merge into into master (I don't really want to deal with the merge conflict...)?

[benchmark.zip](https://github.com/prioritizr/prioritizr/files/6448823/benchmark.zip)

ricschuster commented:
@jeffreyhanson, looks great!
Only one small change made. Otherwise, I think its ready for a new PR and merge into master.
Thanks!
